File: src/futu.js
const {WebSocketServer} = require('ws')
const pty = require('node-pty')

const {
  STATUS,
  OutputManager
} = require('./common')


class FutuManager {
  #cmd
  #login_account
  #login_pwd_md5
  #lang
  #log_level
  #api_port
  #status
  #supervise
  #retry
  #should_log
  #ws
  #clients
  #ready_to_receive_code
  #resolveReadyToReceiveCode
  #child
  #code
  #output

  constructor (cmd, {
    login_account,
    login_pwd_md5,
    lang,
    log_level,
    api_port,
    server_port,

    // Whether to auto-init the FutuOpenD process
    auto_init = true,

    // Whether to supervise the FutuOpenD process, and restart it if it closes
    supervise = true
  }) {
    this.#cmd = cmd
    this.#login_account = login_account
    this.#login_pwd_md5 = login_pwd_md5
    this.#lang = lang
    this.#log_level = log_level
    this.#api_port = api_port
    this.#status = STATUS.ORIGIN
    this.#supervise = supervise
    this.#retry = parseInt(
      // For testing purposes
      process.env.FUTU_RETRY,
      10
    ) || 0

    this.#should_log = log_level !== 'no'

    this.#ws = new WebSocketServer({port: server_port}, () => {
      this.#log(`WebSocket server is listening on port ${server_port}`)
    })

    this.#clients = []

    this.#ws.on('connection', ws => {
      if (this.#status === STATUS.REQUESTING_VERIFICATION_CODE) {
        this.#send({
          type: 'REQUEST_CODE'
        }, [ws])
      }

      if (this.#status === STATUS.CONNECTED) {
        this.#send({
          type: 'CONNECTED'
        }, [ws])
      }

      this.#clients.push(ws)
      ws.on('error', err => {
        this.#error('ws error:', err)
      })

      ws.on('message', msg => {
        const payload = JSON.parse(msg)
        const {
          type,
          code
        } = payload

        if (type === 'VERIFY_CODE') {
          this.verify_code(code)
          return
        }

        if (type === 'INIT') {
          this.#init()
          return
        }

        if (type === 'STATUS') {
          this.#send({
            type: 'STATUS',
            status: this.#status
          }, [ws])
          return
        }
      })
    })

    this.#reset_ready_to_receive_code()

    if (auto_init) {
      this.#init()
    }
  }

  #log(...msg) {
    if (this.#should_log) {
      console.log('[INFO]', ...msg)
    }
  }

  #error(...msg) {
    if (this.#should_log) {
      console.error('[ERROR]', ...msg)
    }
  }

  #reset_ready_to_receive_code() {
    this.#ready_to_receive_code = new Promise((resolve, reject) => {
      this.#resolveReadyToReceiveCode = resolve
    })
  }

  #init() {
    if (this.#status >= STATUS.INIT) {
      // Already inited
      return
    }

    this.#status = STATUS.INIT

    this.#log('Initializing FutuOpenD with options ...', {
      login_account: this.#login_account,
      login_pwd_md5: '<hidden>',
      lang: this.#lang,
      log_level: this.#log_level,
      api_port: this.#api_port
    })

    this.#child = pty.spawn(this.#cmd, [
      `-login_account=${this.#login_account}`,
      `-login_pwd_md5=${this.#login_pwd_md5}`,
      `-lang=${this.#lang}`,
      `-log_level=${this.#log_level}`,
      `-api_port=${this.#api_port}`
    ], {
      name: 'xterm-color',
      cols: 80,
      rows: 30,
      cwd: process.cwd(),
      env: {
        ...process.env,
        FUTU_RETRY: this.#retry
      }
    })

    this.#output = new OutputManager()

    this.#child.on('data', chunk => {
      process.stdout.write(chunk)
      this.#output.add(chunk)

      if (this.#output.includes('req_phone_verify_code')) {
        this.#send({
          type: 'REQUEST_CODE'
        })
        this.#status = STATUS.REQUESTING_VERIFICATION_CODE
        this.#resolveReadyToReceiveCode()
        return
      }

      if (this.#output.includes('Login successful')) {
        this.#send({
          type: 'CONNECTED'
        })
        this.#status = STATUS.CONNECTED
        this.#output.close()
      }
    })

    this.#child.on('error', err => {
      this.#error('FutuOpenD error:', err)
    })

    this.#child.on('exit', (code, signal) => {
      this.#error('FutuOpenD exited')
    })

    this.#child.on('close', () => {
      this.#log('FutuOpenD closed')

      this.#status = STATUS.CLOSED
      this.#send({
        type: 'CLOSED'
      })

      this.#reset_ready_to_receive_code()

      if (this.#supervise) {
        this.#retry++
        this.#init()
      }
    })
  }

  // Send msg to specific clients or all clients
  #send(msg, clients) {
    if (msg.type === 'REQUEST_CODE' && this.#code) {
      // Already has a code
      return
    }

    (clients || this.#clients).forEach(client => {
      client.send(JSON.stringify(msg))
    })
  }

  verify_code(code) {
    this.#code = code

    if (this.#status === STATUS.REQUESTING_VERIFICATION_CODE) {
      this.#set_verify_code()
      return
    }

    if (this.#status === STATUS.CONNECTED) {
      // Already connected, no need to verify code
      return
    }

    this.#ready_to_receive_code.then(() => {
      this.#set_verify_code()
    })
  }

  #set_verify_code() {
    const code = this.#code
    this.#code = undefined

    // this.#ready.then might be called multiple times,
    //   so we need to test the current status again
    if (this.#status !== STATUS.REQUESTING_VERIFICATION_CODE) {
      return
    }

    this.#status = STATUS.VERIFIYING_CODE
    this.#child.write(`input_phone_verify_code -code=${code}\r`)
  }
}


module.exports = {
  FutuManager,
  STATUS
}

File: src/futu.py
import asyncio
import json
import logging
import subprocess
from enum import IntEnum
from typing import (
    List,
    Optional,
    Set
)
import websockets

class Status(IntEnum):
    ORIGIN = -1
    INIT = 0
    REQUESTING_VERIFICATION_CODE = 1
    VERIFYING_CODE = 2
    CONNECTED = 3

class FutuManager:
    _process: Optional[subprocess.Popen]
    _code_future: Optional[asyncio.Future]
    _clients: Set[websockets.WebSocketServerProtocol]

    def __init__(
        self,
        cmd: str,
        login_account: str,
        login_pwd_md5: str,
        login_region: str,
        lang: str,
        log_level: str,
        api_port: int,
        server_port: int,
        auto_init: bool = True
    ):
        self._cmd = cmd
        self._login_account = login_account
        self._login_pwd_md5 = login_pwd_md5
        self._login_region = login_region
        self._lang = lang
        self._log_level = log_level
        self._api_port = api_port
        self._status = Status.ORIGIN
        self._server_port = server_port

        # Initialize logging
        logging.basicConfig(level=logging.INFO)
        self.logger = logging.getLogger('FutuOpenD')

        self._clients = set()
        self._process = None
        self._code_future = None
        self._reset_ready_to_receive_code()

        if auto_init:
            asyncio.create_task(self._init())

    def _reset_ready_to_receive_code(self):
        self._code_future = asyncio.Future()

    async def start_server(self):
        async with websockets.serve(self._handle_connection, 'localhost', self._server_port):
            await asyncio.Future()  # run forever

    async def _handle_connection(self, websocket):
        try:
            self._clients.add(websocket)

            if self._status == Status.REQUESTING_VERIFICATION_CODE:
                await self._send({'type': 'REQUEST_CODE'}, [websocket])

            if self._status == Status.CONNECTED:
                await self._send({'type': 'CONNECTED'}, [websocket])

            async for message in websocket:
                try:
                    payload = json.loads(message)
                    await self._handle_message(payload, websocket)
                except json.JSONDecodeError:
                    self.logger.error('Invalid JSON received')
                    continue

        except websockets.exceptions.ConnectionClosed:
            self.logger.info('Client disconnected')
        finally:
            self._clients.remove(websocket)

    async def _handle_message(self, payload: dict, websocket):
        message_type = payload.get('type')

        if message_type == 'VERIFY_CODE':
            await self.verify_code(payload.get('code'))
        elif message_type == 'INIT':
            await self._init()
        elif message_type == 'STATUS':
            await self._send({
                'type': 'STATUS',
                'status': self._status
            }, [websocket])

    async def _init(self):
        if self._status >= Status.INIT:
            return

        self._status = Status.INIT

        cmd_args = [
            self._cmd,
            f'-login_account={self._login_account}',
            f'-login_pwd_md5={self._login_pwd_md5}',
            f'-login_region={self._login_region}',
            f'-lang={self._lang}',
            f'-log_level={self._log_level}',
            f'-api_port={self._api_port}'
        ]

        self._process = subprocess.Popen(
            cmd_args,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            stdin=subprocess.PIPE,
            text=True,
            bufsize=1
        )

        # Start reading stdout and stderr
        asyncio.create_task(self._read_stdout())
        asyncio.create_task(self._read_stderr())

    async def _read_stdout(self):
        assert self._process is not None

        while True:
            line = await asyncio.get_event_loop().run_in_executor(
                None, self._process.stdout.readline
            )
            if not line:
                break

            self.logger.info(f'stdout: {line.strip()}')

            if 'req_phone_verify_code' in line:
                await self._send({'type': 'REQUEST_CODE'})
                self._status = Status.REQUESTING_VERIFICATION_CODE
                self._code_future.set_result(None)

            elif 'Login successful' in line:
                await self._send({'type': 'CONNECTED'})
                self._status = Status.CONNECTED

    async def _read_stderr(self):
        assert self._process is not None
        while True:
            line = await asyncio.get_event_loop().run_in_executor(
                None, self._process.stderr.readline
            )
            if not line:
                break
            self.logger.error(f'stderr: {line.strip()}')

    async def _send(
        self,
        msg: dict,
        clients: Optional[List[websockets.WebSocketServerProtocol]] = None
    ):
        if clients is None:
            clients = self._clients

        message = json.dumps(msg)
        tasks = [client.send(message) for client in clients]
        if tasks:
            await asyncio.gather(*tasks)

    async def verify_code(self, code: str):
        if self._status == Status.REQUESTING_VERIFICATION_CODE:
            await self._set_verify_code(code)
            return

        if self._status == Status.CONNECTED:
            return

        await self._code_future
        await self._set_verify_code(code)

    async def _set_verify_code(self, code: str):
        if self._status != Status.REQUESTING_VERIFICATION_CODE:
            return

        self._status = Status.VERIFYING_CODE
        assert self._process is not None
        self._process.stdin.write(f'input_phone_verify_code -code={code}\n')
        self._process.stdin.flush()
        self._process.stdin.close()
